larpandoracontent/LArReclustering/ThreeDMultiReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,360 @@
+    {
+        std::string reclusterListName;
+        const ClusterList *pReclusterList {nullptr};
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=,
+                                 PandoraContentApi::RunClusteringAlgorithm(*this, clusteringAlg, pReclusterList, reclusterListName));
+        if (!pReclusterList || pReclusterList->empty())
+            continue;
+
+        float reclusterFom;
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, m_pFomAlgTool->CalcClusteringFom(*pReclusterList, reclusterFom));
+        if (reclusterFom > bestReclusterFom)
+        {
+            bestReclusterFom = reclusterFom;
+            bestReclusterListName = reclusterListName;
+            bestReclusterList = *pReclusterList;
+        }
+    }
+
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::TemporarilyReplaceCurrentList<Cluster>(*this, bestReclusterListName));
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::EndReclustering(*this, bestReclusterListName));
+
+    if (bestReclusterListName == originalClusterListName) // Return the clusters to the original pfos as if nothing happened...
+    {
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, AddClustersToPfos(pfoToFreeClusters));
+        return STATUS_CODE_SUCCESS;
+    }
+    else
+    {
+        // Delete the original pfos in preparation for making new ones
+        // ATTN This will break any existing particle hierarchy
+        for (const Pfo *const pPfo : pfosToRecluster)
+            PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Delete(*this, pPfo, m_pfoListName));
+    }
+
+    // The 3D clusters are reclustered, now need to manually recluster the 2D clusters to be consistent with the 3D reclustering
+    // First, remove the original 2D clusters
+    std::map<HitType, CaloHitList> viewToFreeCaloHits2D = { {TPC_VIEW_U, {}}, {TPC_VIEW_V, {}}, {TPC_VIEW_W, {}} };
+    for (const auto &[view, freeClusters] : viewToFreeClusters)
+    {
+        if (view == TPC_3D)
+            continue;
+        PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, FreeCaloHitsFromClusters(freeClusters, view, viewToFreeCaloHits2D.at(view)));
+    }
+
+    // Now make new 2D clusters following the new 3D clusters, and make new corresponding pfos
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, BuildNewPfos(bestReclusterList, viewToFreeCaloHits2D));
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDMultiReclusteringAlgorithm::FreeClustersFromPfos(const PfoList &pfos,
+                                                                  std::map<HitType, ClusterList> &viewToFreeClusters,
+                                                                  std::map<const Pfo *const, ClusterList> &pfoToFreeClusters)
+{
+    for (const Pfo *const pPfo : pfos)
+    {
```

Comment:
If your logic of your alg (and I dont know if it does) always ignores pfos with no 3D hits, you could just skip over them at this point. Theres a function in PfoHelper that returns this info for you.

Reply:
I think the best thing to do is leave it to the figure of merit tool `GetPfosToRecluster` to tell us to recluster only pfos with 3D hits if that is desired. There may be some weird use case where you have a pfo with pretty bad 2D clusters and no 3D cluster which you want to throw into the reclustering with other pfos that do have 3D hits so that its 2D clusters can be reclustered according to the new 3D clustering... If that makes sense :grin: